src/qml_benchmarks/models/base.py

```diff
@@ -0,0 +1,299 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base classes for models."""
+
+import copy
+from abc import abstractmethod
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+import optax
+from qml_benchmarks.model_utils import train
+from sklearn.base import BaseEstimator
+
+
+class BaseGenerator(BaseEstimator):
+    """
+    A base class for generative models.
+
+    We use Scikit-learn's `BaseEstimator` so that we can take advantage of
+    Scikit-learn's hyperparameter search algorithms like `GridSearchCV` or
+    `RandomizedSearchCV` for hyperparameter tuning.
+
+    Args:
+        dim:
+            The dimensionality of the samples, e.g., for spins it could be an
+            integer or a tuple specifying a grid.
+    """
+
+    def __init__(self, dim: int or tuple[int]) -> None:
+        self.dim = dim
+
+    @abstractmethod
+    def initialize(self, x: any = None):
+        """
+        Initialize the model and create the model parameters.
+
+        Args:
+            x: An example data or dimensionality of the model parameters.
+        """
+        # self.dim = x.shape[1:]
+        pass
+
+    @abstractmethod
+    def sample(self, num_samples: int) -> any:
+        """
+        Sample from the model.
+
+        Args:
+            num_samples: The number of samples to generate.
+        """
+        pass
+
+
+class EnergyBasedModel(BaseGenerator):
+    """
+    A base class for energy-based generative models with common functionalities.
+
+    We use Scikit-learn's `BaseEstimator` so that we can take advantage of
+    Scikit-learn's hyperparameter search algorithms like `GridSearchCV` or
+    `RandomizedSearchCV` for hyperparameter tuning.
+
+    The parameters of the model are stored in the `params_` attribute. The
+    model hyperparameters are explicitly passed to the constructor. See the
+    `BaseEstimator` documentation for more details.
+
+    References:
+    Teh, Yee Whye, Max Welling, Simon Osindero, and Geoffrey E. Hinton.
+    "Energy-Based Models for Sparse Overcomplete Representations."
+    Journal of Machine Learning Research, vol. 4, 2003, pp. 1235-1260.
+    """
+
+    def __init__(
+        self,
+        dim: int = None,
+        learning_rate=0.001,
+        batch_size=32,
+        max_steps=10000,
+        cdiv_steps=100,
+        convergence_interval=None,
+        random_state=42,
+        jit=True,
```

Comment:
these init arguments weren't appearing as options in my Jupyter notebook when pressing tab. Do you know if there is a way to make that work?